File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.scala
object Solution {
    def removeNthFromEnd(head: ListNode, n: Int): ListNode = {
        var d = new ListNode(0)
        d.next = head

        var fast = d
        for(i <- 1 to n) {
            fast = fast.next
        }

        var slow = d
        var prev: ListNode = null

        while(fast != null) {
            prev = slow
            slow = slow.next
            fast = fast.next
        }

        prev.next = slow.next
        d.next
    }
}

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode root = new ListNode();
        root.next = head;
        
        ListNode fast = root;
        ListNode slow = root;
        ListNode prev = null;
        
        for(int i = 0; i < n; i++, fast = fast.next);
        
        while(fast != null) {
            fast = fast.next;
            prev = slow;
            slow = slow.next;
        }
        
        prev.next = slow.next;
        
        return root.next;
    }
}